### ISSUE TITLE
Sik-Ho Tang | Review: Highway Networks -- Gating Function To Highway (Image Classification).

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#91]
Sik-Ho Tang. [Review: Highway Networks — Gating Function To Highway (Image Classification)](https://towardsdatascience.com/review-highway-networks-gating-function-to-highway-image-classification-5a33833797b5).

**Comment by NorbertZheng** [comment]
## Overview
Inthis story, Highway Networks is briefly presented. This is a 2015 work. At that moment, it is found that
 - there is difficulties optimizing a very deep neural network.

However, it’s still an open problem why it is difficult to optimize a deep network. (Of course, later, it is probably due to gradient vanishing problem.) Inspired by **Long Short-Term Memory (LSTM)**, authors thereby
 - make use of gating function to adaptively transform or bypass the signal so that the network can go deeper.

A deep network with more than 1000 layers can also be optimized. I choose to present this paper so that I can introduce the gating function.

Highway Networks initially was presented in 2015 ICML Deep Learning Workshop and published as a 2015 arXiv tech report with over 600 citations. And later on it is extended and published in 2015 NIPS with over 500 citations.

**Comment by NorbertZheng** [comment]
## Highway Networks
### Plain Network
Before talking about Highway Networks, Let’s start with plain network which consists of $L$ layers where the $l$-th layer (with omitting the symbol for layer):
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/f9d37618-5cc4-4f82-9774-563c4000bd25)

where $x$ is input, $W_{H}$ is the weight, $H$ is the transform function followed by an activation function and y is the output. And for $i$-th unit:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/2782687a-c397-4c29-bec4-4a58714c55b0)

We compute the $y_{i}$ and pass it to next layer.

**Comment by NorbertZheng** [comment]
### Highway Network
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/860df7ef-3cb5-443e-9810-eb613dd37dc4)
Highway Circuit.

In highway network, two non-linear transforms $T$ and $C$ are introduced:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/d06e51ea-2062-46cd-a7a6-bcb904ee186a)

where $T$ is the Transform Gate and $C$ is the Carry Gate.

In particular, $C = 1 - T$:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/7f59d56e-ae45-4300-8e10-b8c3690ff0cd)

We can have below conditions for particular $T$ values:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/1bc9ce70-982f-4889-91ed-aadefbec98be)

 - When $T=0$, we pass the input as output directly which creates an information highway. That’s why it is called Highway Network !!!
 - When $T=1$, we use the non-linear activated transformed input as output.

Here, in contrast to the $i$-th unit in plain network, authors introduce the block concept. For $i$-th block, there is a block state $H_{i}(x)$, and and transform gate output $T_{i}(x)$. And the corresponding block output $y_{i}$:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/c93af7df-fb19-4b20-a7b9-dc8f9061b9cd)

which is connected to the next layer.

Formally, $T(x)$ is the sigmoid function:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/d585b94a-6087-4a40-82a5-4cf692901194)

If we remember, sigmoid function caps the output between 0 to 1. When the input has too small value, it becomes 0. When the input has too large value, it becomes 1. Therefore, by learning $W_{T}$ and $b_{T}$, the network can adaptively pass $H(x)$ or just pass $x$ to next layer.

 - And author claims that this sigmoid function format helps to have a simple initialization scheme for $W_{T}$ which is independent of the nature of $H$.

$b_{T}$ can be initialized with a negative value (e.g. -1, -3 etc.) such that the network is initially biased towards carry behavior.

The above idea is inspired by LSTM as authors mentioned. (LSTM is a very famous module mainly used in Natural Language Processing (NLP)).

And Stochastic Gradient Descent (SGD) did not stall for networks with more than 1000 layers. However, the exact results has not been provided.

**Comment by NorbertZheng** [comment]
## Results
### MNIST
The first layer is a fully connected plain layer followed by 9, 19, 49, or 99 fully connected plain or highway layers. Finally, the network output is produced by a softmax layer.

All networks are thin: **each layer has 50 blocks for highway networks and 71 units for plain networks**, yielding roughly identical numbers of parameters (~5000) per layer.
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/1aad7dcb-1163-4c64-b7e1-2d74f922a79b)
The best training curves for Plain Networks (Left) The best training curves for Highway Networks (Right) Mean performance of top 10 (out of 100) hyperparameter settings.

- The error is evaluated on the training set, instead of the test set!!!
- **Only deeper networks are better with Highway Blocks!!!**

~~As shown above, the errors obtained by Highway Networks are always smaller than those by Plain Network.~~

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/e5927be4-fec6-4f4c-8a7c-487b400cda14)
MNIST Test Accuracy.

10-layer convolutional highway networks on MNIST are trained, using two architectures, each with 9 convolutional layers followed by a softmax output. The number of filter maps (width) was set to 16 and 32 for all the layers.

Compared with Maxout and DSN, Highway Networks obtained similar accuracy but with much fewer number of parameters. (If interested, please visit my review on NoC for a very brief introduction of Maxout.)

**Comment by NorbertZheng** [comment]
### CIFAR-10 & CIFAR-100
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/bda8c4f6-66ba-4a27-8ce2-82b1c686581a)
CIFAR-10 Test Accuracy.

Fitnet cannot optimize the networks directly when the networks are deep. It needs two-stage training.

By using gating function, Highway can optimize the deep networks directly. In particular, Highway B obtains highest accuracy with 19 layers.

Though Highway C is inferior to Highway B, it stills can be optimized directly due to the existence of gating function.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/3f61a5c0-fb73-4959-a5b8-f612012ad6a3)
CIFAR-10, CIFAR-100 Test Accuracy.

Here, the fully connected layer used in the networks in the previous experiment is replaced with a convolutional layer with a receptive field of size one and a global average pooling layer.

Highway Network can obtain comparable performance on CIFAR-10 and highest accuracy on CIFAR-100.

**Comment by NorbertZheng** [comment]
## Analyses
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/4d28b096-de92-40dd-9142-058853fde2ee)
Visualization of best 50 hidden-layer highway networks trained on MNIST (top row) and CIFAR-100 (bottom row).

The above figure shows the bias, the mean activity over all training samples, and the activity for a single random sample for each transform gate respectively. Block outputs for the same single sample are displayed in the last column.

For the CIFAR-100 network, the biases increase with depth forming a gradient.
 - **The strong negative biases at low depths are not used to shut down the gates, but to make them more selective.** This makes the transform gate activity for a single example (column 3) is very sparse.

For the CIFAR-100 case, most transform gates are active on average, while they show very selective activity for the single example. This implies that for each sample only a few blocks perform transformation but different blocks are utilized by different samples.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/4c7d3b10-fc41-45c2-8323-c667d855121f)
Visualization showing the extent to which the mean transform gate activity for certain classes differs from the mean activity over all training samples.

For MNIST digits 0 and 7, substantial differences can be seen within the first 15 layers.

For CIFAR class numbers 0 and 1, the differences are sparser and spread out over all layers.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/0e47bfee-79b7-4d58-b865-4f2aa5782cfe)
Lesioned training set performance.

By lesioning, It is meant to manually set all the transform gates of a layer to 0 forcing it to simply copy its inputs. As shown above, for each layer, the network is evaluated on the full training set with the gates of that layer closed.

For MNIST (left) it can be seen that the error rises significantly if any one of the early layers is removed, but layers 15-45 seem to have close to no effect on the final performance. About 60% of the layers don’t learn to contribute to the final result, likely because
 - **MNIST is a simple dataset that doesn’t require much depth.**

While CIFAR-10 is a relatively complex dataset, the error rises more significantly.

**Comment by NorbertZheng** [comment]
## References
 - [2015] [arXiv] [Highway Networks](https://arxiv.org/abs/1505.00387).
 - [2015] [NIPS] [Training Very Deep Networks](https://arxiv.org/abs/1507.06228).
